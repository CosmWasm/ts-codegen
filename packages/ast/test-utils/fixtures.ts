// @ts-nocheck
import { JSONSchema } from '../src/types';

import canExecuteRelayResponseData from '../../../__fixtures__/vectis/govec/can_execute_relay_response.json';
import cosmosMsgForEmptyData from '../../../__fixtures__/vectis/govec/cosmos_msg_for__empty.json';
import executeMsgForEmptyData from '../../../__fixtures__/vectis/govec/execute_msg_for__empty.json';
import infoResponseData from '../../../__fixtures__/vectis/govec/info_response.json';
import relayTransactionData from '../../../__fixtures__/vectis/govec/relay_transaction.json';

import nftIdlVersionData from '../../../__fixtures__/idl-version/accounts-nft/account-nft.json';

export const nftIdlVersion: JSONSchema = nftIdlVersionData;
export const vectisCanExecuteRelayResponse: JSONSchema = canExecuteRelayResponseData;
export const vectisCosmosMsgForEmpty: JSONSchema = cosmosMsgForEmptyData;
export const vectisExecuteMsgForEmpty: JSONSchema = executeMsgForEmptyData;
export const vectisInfoResponse: JSONSchema = infoResponseData;
export const vectisRelayTransaction: JSONSchema = relayTransactionData;
