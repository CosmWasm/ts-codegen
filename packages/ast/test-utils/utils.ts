import generate from '@babel/generator';
import { RenderOptions } from '@cosmwasm/ts-codegen-ast/types';
import { ExecuteMsg, IDLObject, JSONSchema, QueryMsg } from '@cosmwasm/ts-codegen-types';
import { readFileSync } from 'fs';
import { sync as glob } from 'glob';
import { join } from 'path';

import { RenderContext } from '../src/context';

export const expectCode = (ast: any): void => {
  expect(
    generate(ast).code
  ).toMatchSnapshot();
};

export const printCode = (ast: any): void => {
  console.log(
    generate(ast).code
  );
};

export const makeContext = (
  schema: JSONSchema,
  options?: RenderOptions,
  responses?: Record<string, JSONSchema>
) => {
  return new RenderContext({
    schemas: [schema],
    responses
  }, options);
};

interface GlobContractLegacy {
  name: `/${string}.json`;
  content: JSONSchema;
}

interface GlobContract {
  name: `/${string}.json`;
  content: IDLObject;
}

const fixtureDir = join(__dirname, '../../../__fixtures__');
const globCache: Record<string, GlobContractLegacy[] | GlobContract[]> = {};

export const globIdlBasedContracts = (p: string): GlobContract[] => {
  if (globCache[p]) return globCache[p] as GlobContract[];
  // @ts-ignore
  const contracts: GlobContract[] = glob(join(fixtureDir, p, '/*.json')).map(file => {
    return {
      name: file.split(join('__fixtures__', p))[1],
      content: JSON.parse(readFileSync(file, 'utf-8'))
    };
  });
  globCache[p] = contracts;
  return contracts;
};

export const globLegacyContracts = (p: string): GlobContractLegacy[] => {
  if (globCache[p]) return globCache[p] as GlobContractLegacy[];
  // @ts-ignore
  const contracts: GlobContractLegacy[] = glob(join(fixtureDir, p, '/*.json')).map(file => {
    return {
      name: file.split(join('__fixtures__', p))[1],
      content: JSON.parse(readFileSync(file, 'utf-8'))
    };
  });
  globCache[p] = contracts;
  return contracts;
};

export const getMsgExecuteLegacyFixture = (
  p: string,
  name: `/${string}.json`
): ExecuteMsg => {
  const contracts = globLegacyContracts(p);
  const contract = contracts.find(a => a.name === name);
  return contract.content as ExecuteMsg;
};

export const getMsgQueryLegacyFixture = (
  p: string,
  name: `/${string}.json`
): QueryMsg => {
  const contracts = globLegacyContracts(p);
  const contract = contracts.find(a => a.name === name);
  return contract.content as QueryMsg;
};

export const getLegacyFixture = (
  p: string,
  name: `/${string}.json`
): JSONSchema => {
  const contracts = globLegacyContracts(p);
  const contract = contracts.find(a => a.name === name);
  return contract.content as JSONSchema;
};
