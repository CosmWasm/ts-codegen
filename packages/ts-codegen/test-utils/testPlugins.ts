import {
  BuilderContext,
  ContractInfo,
  RenderContext,
  RenderContextBase,
  RenderOptions,
} from "wasm-ast-types";
import { ClientPlugin, MessageComposerPlugin, TSBuilderOptions } from "../src";
import { ContractsContextProviderPlugin } from "../src/plugins/provider";
import { ContractsProviderBundlePlugin } from "../src/plugins/provider-bundle";

export class TestClientPlugin extends ClientPlugin {
  builderContext: BuilderContext;

  constructor(opts: RenderOptions, builderContext: BuilderContext) {
    super(opts);
    this.builderContext = builderContext;
  }

  initContext(
    contract: ContractInfo,
    options?: RenderOptions
  ): RenderContextBase<RenderOptions> {
    return new RenderContext(contract, options, this.builderContext);
  }
}

export class TestMessageComposerPlugin extends MessageComposerPlugin {
  builderContext: BuilderContext;

  constructor(opts: RenderOptions, builderContext: BuilderContext) {
    super(opts);
    this.builderContext = builderContext;
  }

  initContext(
    contract: ContractInfo,
    options?: RenderOptions
  ): RenderContextBase<RenderOptions> {
    return new RenderContext(contract, options, this.builderContext);
  }
}

export class TestContractsContextProviderPlugin extends ContractsContextProviderPlugin {
  builderContext: BuilderContext;

  constructor(opts: TSBuilderOptions, builderContext: BuilderContext) {
    super(opts);
    this.builderContext = builderContext;
  }

  initContext(
    contract: ContractInfo,
    options?: TSBuilderOptions
  ): RenderContextBase<TSBuilderOptions> {
    return new RenderContext(contract, options, this.builderContext);
  }
}

export class TestContractsProviderBundlePlugin extends ContractsProviderBundlePlugin {
  builderContext: BuilderContext;

  constructor(opts: TSBuilderOptions, builderContext: BuilderContext) {
    super(opts);
    this.builderContext = builderContext;
  }

  initContext(
    contract: ContractInfo,
    options?: TSBuilderOptions
  ): RenderContextBase<TSBuilderOptions> {
    return new RenderContext(contract, options, this.builderContext);
  }
}
