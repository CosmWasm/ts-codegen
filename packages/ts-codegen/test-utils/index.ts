import {
  BuilderContext,
  ContractInfo,
  ReactQueryOptions,
  RenderOptions,
  TSTypesOptions,
} from "wasm-ast-types";
import { ReactQueryPlugin } from "../src/plugins/react-query";
import { TypesPlugin } from "../src/plugins/types";
import { ClientPlugin } from "../src/plugins/client";
import { MessageComposerPlugin } from "../src/plugins/message-composer";
import { RecoilPlugin } from "../src/plugins/recoil";
import { MessageBuilderPlugin } from "../src/plugins/message-builder";
import { ContractsContextProviderPlugin } from "../src/plugins/provider";
import { ContractsProviderBundlePlugin } from "../src/plugins/provider-bundle";
import { TSBuilderOptions } from "../src/builder";
import { createDefaultContractInfo } from "../src/utils/contracts";
import {
  TestContractsContextProviderPlugin,
  TestClientPlugin,
  TestMessageComposerPlugin,
  TestContractsProviderBundlePlugin,
} from "./testPlugins";
import deepmerge from "deepmerge";

export const testDefaultOptions: RenderOptions = {
  enabled: true,
  types: {
    enabled: true,
    itemsUseTuples: false,
    aliasExecuteMsg: false,
  },
  client: {
    enabled: true,
    execExtendsQuery: true,
    noImplicitOverride: false,
  },
  recoil: {
    enabled: true,
  },
  messageComposer: {
    enabled: true,
  },
  messageBuilder: {
    enabled: true,
  },
  reactQuery: {
    enabled: true,
    optionalClient: false,
    version: "v4",
    mutations: false,
    camelize: true,
    queryKeys: false,
  },
  useContractsHook: {
    enabled: true,
  },
};

export async function generateReactQuery(
  contractName: string,
  contractInfo: ContractInfo,
  outPath: string,
  opts?: ReactQueryOptions
) {
  const options = opts
    ? deepmerge(testDefaultOptions, { reactQuery: opts })
    : testDefaultOptions;

  await new ReactQueryPlugin(options).render(
    outPath,
    contractName,
    contractInfo
  );
}

export async function generateTypes(
  contractName: string,
  contractInfo: ContractInfo,
  outPath: string,
  opts?: TSTypesOptions
) {
  await new TypesPlugin(opts ?? testDefaultOptions).render(
    outPath,
    contractName,
    contractInfo
  );
}

export async function generateClient(
  contractName: string,
  contractInfo: ContractInfo,
  outPath: string,
  builderContext?: BuilderContext,
  opts?: RenderOptions
) {
  if (builderContext) {
    await new TestClientPlugin(
      opts ?? testDefaultOptions,
      builderContext
    ).render(outPath, contractName, contractInfo);
  } else {
    await new ClientPlugin(opts ?? testDefaultOptions).render(
      outPath,
      contractName,
      contractInfo
    );
  }
}

export async function generateMessageComposer(
  contractName: string,
  contractInfo: ContractInfo,
  outPath: string,
  builderContext?: BuilderContext,
  opts?: RenderOptions
) {
  if (builderContext) {
    await new TestMessageComposerPlugin(
      opts ?? testDefaultOptions,
      builderContext
    ).render(outPath, contractName, contractInfo);
  } else {
    await new MessageComposerPlugin(opts ?? testDefaultOptions).render(
      outPath,
      contractName,
      contractInfo
    );
  }
}

export async function generateMessageBuilder(
  contractName: string,
  contractInfo: ContractInfo,
  outPath: string,
  opts?: RenderOptions
) {
  await new MessageBuilderPlugin(opts ?? testDefaultOptions).render(
    outPath,
    contractName,
    contractInfo
  );
}

export async function generateRecoil(
  contractName: string,
  contractInfo: ContractInfo,
  outPath: string,
  opts?: RenderOptions
) {
  await new RecoilPlugin(opts ?? testDefaultOptions).render(
    outPath,
    contractName,
    contractInfo
  );
}

export async function generateContractHooks(
  contractName: string,
  contractInfo: ContractInfo,
  outPath: string,
  builderContext?: BuilderContext
) {
  await new TestContractsContextProviderPlugin(
    {
      enabled: true,
      useShorthandCtor: true,
      useContractsHook: {
        enabled: true,
      },
    },
    builderContext
  ).render(outPath, contractName, contractInfo);

  await new TestContractsProviderBundlePlugin(
    {
      enabled: true,
      useShorthandCtor: true,
      useContractsHook: {
        enabled: true,
      },
    },
    builderContext
  ).render(outPath, undefined, contractInfo ?? createDefaultContractInfo());
}
