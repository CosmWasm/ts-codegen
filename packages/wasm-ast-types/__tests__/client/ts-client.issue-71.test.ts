import { globContracts, makeContext } from '../../../test-utils'
import {
    createQueryClass,
    createExecuteClass,
    createExecuteInterface,
    createTypeInterface
} from '../client'
import { expectCode } from '../../../test-utils';
import cases from 'jest-in-case';

const contracts = globContracts('issues/71');

cases('execute_msg_for__empty', async opts => {
    const ctx = makeContext(opts.content);
    expectCode(createTypeInterface(
        ctx,
        opts.content
    ));
}, contracts);

cases('query classes', async opts => {
    const ctx = makeContext(opts.content);
    expectCode(createQueryClass(
        ctx,
        'SG721QueryClient',
        'SG721ReadOnlyInstance',
        opts.content
    ))
}, contracts);

cases('execute class', async opts => {
    const ctx = makeContext(opts.content);
    expectCode(createExecuteClass(
        ctx,
        'SG721Client',
        'SG721Instance',
        null,
        opts.content
    ))
}, contracts);

cases('execute interface', async opts => {
    const ctx = makeContext(opts.content);
    expectCode(createExecuteInterface(
        ctx,
        'SG721Instance',
        null,
        opts.content
    ))
}, contracts);

