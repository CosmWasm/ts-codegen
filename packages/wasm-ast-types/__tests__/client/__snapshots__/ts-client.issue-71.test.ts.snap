// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`execute class /issues/71/execute_msg.json 1`] = `
"export class SG721Client implements SG721Instance {
  client: SigningCosmWasmClient;
  sender: string;
  contractAddress: string;

  constructor(client: SigningCosmWasmClient, sender: string, contractAddress: string) {
    this.client = client;
    this.sender = sender;
    this.contractAddress = contractAddress;
    this.receive = this.receive.bind(this);
    this.updateConfig = this.updateConfig.bind(this);
    this.updatePoolConfig = this.updatePoolConfig.bind(this);
    this.addToRegistery = this.addToRegistery.bind(this);
    this.createPoolInstance = this.createPoolInstance.bind(this);
    this.joinPool = this.joinPool.bind(this);
    this.swap = this.swap.bind(this);
    this.proposeNewOwner = this.proposeNewOwner.bind(this);
    this.dropOwnershipProposal = this.dropOwnershipProposal.bind(this);
    this.claimOwnership = this.claimOwnership.bind(this);
  }

  receive = async ({
    amount,
    msg,
    sender
  }: {
    amount: Uint128;
    msg: Binary;
    sender: string;
  }, fee: number | StdFee | \\"auto\\" = \\"auto\\", memo?: string, _funds?: Coin[]): Promise<ExecuteResult> => {
    return await this.client.execute(this.sender, this.contractAddress, {
      receive: {
        amount,
        msg,
        sender
      }
    }, fee, memo, _funds);
  };
  updateConfig = async ({
    feeCollector,
    generatorAddress,
    lpTokenCodeId
  }: {
    feeCollector?: string;
    generatorAddress?: string;
    lpTokenCodeId?: number;
  }, fee: number | StdFee | \\"auto\\" = \\"auto\\", memo?: string, _funds?: Coin[]): Promise<ExecuteResult> => {
    return await this.client.execute(this.sender, this.contractAddress, {
      update_config: {
        fee_collector: feeCollector,
        generator_address: generatorAddress,
        lp_token_code_id: lpTokenCodeId
      }
    }, fee, memo, _funds);
  };
  updatePoolConfig = async ({
    isDisabled,
    newFeeInfo,
    poolType
  }: {
    isDisabled?: boolean;
    newFeeInfo?: FeeInfo;
    poolType: PoolType;
  }, fee: number | StdFee | \\"auto\\" = \\"auto\\", memo?: string, _funds?: Coin[]): Promise<ExecuteResult> => {
    return await this.client.execute(this.sender, this.contractAddress, {
      update_pool_config: {
        is_disabled: isDisabled,
        new_fee_info: newFeeInfo,
        pool_type: poolType
      }
    }, fee, memo, _funds);
  };
  addToRegistery = async ({
    newPoolConfig
  }: {
    newPoolConfig: PoolConfig;
  }, fee: number | StdFee | \\"auto\\" = \\"auto\\", memo?: string, _funds?: Coin[]): Promise<ExecuteResult> => {
    return await this.client.execute(this.sender, this.contractAddress, {
      add_to_registery: {
        new_pool_config: newPoolConfig
      }
    }, fee, memo, _funds);
  };
  createPoolInstance = async ({
    assetInfos,
    initParams,
    lpTokenName,
    lpTokenSymbol,
    poolType
  }: {
    assetInfos: AssetInfo[];
    initParams?: Binary;
    lpTokenName?: string;
    lpTokenSymbol?: string;
    poolType: PoolType;
  }, fee: number | StdFee | \\"auto\\" = \\"auto\\", memo?: string, _funds?: Coin[]): Promise<ExecuteResult> => {
    return await this.client.execute(this.sender, this.contractAddress, {
      create_pool_instance: {
        asset_infos: assetInfos,
        init_params: initParams,
        lp_token_name: lpTokenName,
        lp_token_symbol: lpTokenSymbol,
        pool_type: poolType
      }
    }, fee, memo, _funds);
  };
  joinPool = async ({
    assets,
    autoStake,
    lpToMint,
    poolId,
    recipient,
    slippageTolerance
  }: {
    assets?: Asset[];
    autoStake?: boolean;
    lpToMint?: Uint128;
    poolId: Uint128;
    recipient?: string;
    slippageTolerance?: Decimal;
  }, fee: number | StdFee | \\"auto\\" = \\"auto\\", memo?: string, _funds?: Coin[]): Promise<ExecuteResult> => {
    return await this.client.execute(this.sender, this.contractAddress, {
      join_pool: {
        assets,
        auto_stake: autoStake,
        lp_to_mint: lpToMint,
        pool_id: poolId,
        recipient,
        slippage_tolerance: slippageTolerance
      }
    }, fee, memo, _funds);
  };
  swap = async ({
    recipient,
    swapRequest
  }: {
    recipient?: string;
    swapRequest: SingleSwapRequest;
  }, fee: number | StdFee | \\"auto\\" = \\"auto\\", memo?: string, _funds?: Coin[]): Promise<ExecuteResult> => {
    return await this.client.execute(this.sender, this.contractAddress, {
      swap: {
        recipient,
        swap_request: swapRequest
      }
    }, fee, memo, _funds);
  };
  proposeNewOwner = async ({
    expiresIn,
    owner
  }: {
    expiresIn: number;
    owner: string;
  }, fee: number | StdFee | \\"auto\\" = \\"auto\\", memo?: string, _funds?: Coin[]): Promise<ExecuteResult> => {
    return await this.client.execute(this.sender, this.contractAddress, {
      propose_new_owner: {
        expires_in: expiresIn,
        owner
      }
    }, fee, memo, _funds);
  };
  dropOwnershipProposal = async (fee: number | StdFee | \\"auto\\" = \\"auto\\", memo?: string, _funds?: Coin[]): Promise<ExecuteResult> => {
    return await this.client.execute(this.sender, this.contractAddress, {
      drop_ownership_proposal: {}
    }, fee, memo, _funds);
  };
  claimOwnership = async (fee: number | StdFee | \\"auto\\" = \\"auto\\", memo?: string, _funds?: Coin[]): Promise<ExecuteResult> => {
    return await this.client.execute(this.sender, this.contractAddress, {
      claim_ownership: {}
    }, fee, memo, _funds);
  };
}"
`;

exports[`execute interface /issues/71/execute_msg.json 1`] = `
"export interface SG721Instance {
  contractAddress: string;
  sender: string;
  receive: ({
    amount,
    msg,
    sender
  }: {
    amount: Uint128;
    msg: Binary;
    sender: string;
  }, fee?: number | StdFee | \\"auto\\", memo?: string, _funds?: Coin[]) => Promise<ExecuteResult>;
  updateConfig: ({
    feeCollector,
    generatorAddress,
    lpTokenCodeId
  }: {
    feeCollector?: string;
    generatorAddress?: string;
    lpTokenCodeId?: number;
  }, fee?: number | StdFee | \\"auto\\", memo?: string, _funds?: Coin[]) => Promise<ExecuteResult>;
  updatePoolConfig: ({
    isDisabled,
    newFeeInfo,
    poolType
  }: {
    isDisabled?: boolean;
    newFeeInfo?: FeeInfo;
    poolType: PoolType;
  }, fee?: number | StdFee | \\"auto\\", memo?: string, _funds?: Coin[]) => Promise<ExecuteResult>;
  addToRegistery: ({
    newPoolConfig
  }: {
    newPoolConfig: PoolConfig;
  }, fee?: number | StdFee | \\"auto\\", memo?: string, _funds?: Coin[]) => Promise<ExecuteResult>;
  createPoolInstance: ({
    assetInfos,
    initParams,
    lpTokenName,
    lpTokenSymbol,
    poolType
  }: {
    assetInfos: AssetInfo[];
    initParams?: Binary;
    lpTokenName?: string;
    lpTokenSymbol?: string;
    poolType: PoolType;
  }, fee?: number | StdFee | \\"auto\\", memo?: string, _funds?: Coin[]) => Promise<ExecuteResult>;
  joinPool: ({
    assets,
    autoStake,
    lpToMint,
    poolId,
    recipient,
    slippageTolerance
  }: {
    assets?: Asset[];
    autoStake?: boolean;
    lpToMint?: Uint128;
    poolId: Uint128;
    recipient?: string;
    slippageTolerance?: Decimal;
  }, fee?: number | StdFee | \\"auto\\", memo?: string, _funds?: Coin[]) => Promise<ExecuteResult>;
  swap: ({
    recipient,
    swapRequest
  }: {
    recipient?: string;
    swapRequest: SingleSwapRequest;
  }, fee?: number | StdFee | \\"auto\\", memo?: string, _funds?: Coin[]) => Promise<ExecuteResult>;
  proposeNewOwner: ({
    expiresIn,
    owner
  }: {
    expiresIn: number;
    owner: string;
  }, fee?: number | StdFee | \\"auto\\", memo?: string, _funds?: Coin[]) => Promise<ExecuteResult>;
  dropOwnershipProposal: (fee?: number | StdFee | \\"auto\\", memo?: string, _funds?: Coin[]) => Promise<ExecuteResult>;
  claimOwnership: (fee?: number | StdFee | \\"auto\\", memo?: string, _funds?: Coin[]) => Promise<ExecuteResult>;
}"
`;

exports[`execute_msg_for__empty /issues/71/execute_msg.json 1`] = `"export type ExecuteMsg = ExecuteMsg;"`;

exports[`query classes /issues/71/execute_msg.json 1`] = `
"export class SG721QueryClient implements SG721ReadOnlyInstance {
  client: CosmWasmClient;
  contractAddress: string;

  constructor(client: CosmWasmClient, contractAddress: string) {
    this.client = client;
    this.contractAddress = contractAddress;
    this.receive = this.receive.bind(this);
    this.updateConfig = this.updateConfig.bind(this);
    this.updatePoolConfig = this.updatePoolConfig.bind(this);
    this.addToRegistery = this.addToRegistery.bind(this);
    this.createPoolInstance = this.createPoolInstance.bind(this);
    this.joinPool = this.joinPool.bind(this);
    this.swap = this.swap.bind(this);
    this.proposeNewOwner = this.proposeNewOwner.bind(this);
    this.dropOwnershipProposal = this.dropOwnershipProposal.bind(this);
    this.claimOwnership = this.claimOwnership.bind(this);
  }

  receive = async ({
    amount,
    msg,
    sender
  }: {
    amount: Uint128;
    msg: Binary;
    sender: string;
  }): Promise<ReceiveResponse> => {
    return this.client.queryContractSmart(this.contractAddress, {
      receive: {
        amount,
        msg,
        sender
      }
    });
  };
  updateConfig = async ({
    feeCollector,
    generatorAddress,
    lpTokenCodeId
  }: {
    feeCollector?: string;
    generatorAddress?: string;
    lpTokenCodeId?: number;
  }): Promise<UpdateConfigResponse> => {
    return this.client.queryContractSmart(this.contractAddress, {
      update_config: {
        fee_collector: feeCollector,
        generator_address: generatorAddress,
        lp_token_code_id: lpTokenCodeId
      }
    });
  };
  updatePoolConfig = async ({
    isDisabled,
    newFeeInfo,
    poolType
  }: {
    isDisabled?: boolean;
    newFeeInfo?: FeeInfo;
    poolType: PoolType;
  }): Promise<UpdatePoolConfigResponse> => {
    return this.client.queryContractSmart(this.contractAddress, {
      update_pool_config: {
        is_disabled: isDisabled,
        new_fee_info: newFeeInfo,
        pool_type: poolType
      }
    });
  };
  addToRegistery = async ({
    newPoolConfig
  }: {
    newPoolConfig: PoolConfig;
  }): Promise<AddToRegisteryResponse> => {
    return this.client.queryContractSmart(this.contractAddress, {
      add_to_registery: {
        new_pool_config: newPoolConfig
      }
    });
  };
  createPoolInstance = async ({
    assetInfos,
    initParams,
    lpTokenName,
    lpTokenSymbol,
    poolType
  }: {
    assetInfos: AssetInfo[];
    initParams?: Binary;
    lpTokenName?: string;
    lpTokenSymbol?: string;
    poolType: PoolType;
  }): Promise<CreatePoolInstanceResponse> => {
    return this.client.queryContractSmart(this.contractAddress, {
      create_pool_instance: {
        asset_infos: assetInfos,
        init_params: initParams,
        lp_token_name: lpTokenName,
        lp_token_symbol: lpTokenSymbol,
        pool_type: poolType
      }
    });
  };
  joinPool = async ({
    assets,
    autoStake,
    lpToMint,
    poolId,
    recipient,
    slippageTolerance
  }: {
    assets?: Asset[];
    autoStake?: boolean;
    lpToMint?: Uint128;
    poolId: Uint128;
    recipient?: string;
    slippageTolerance?: Decimal;
  }): Promise<JoinPoolResponse> => {
    return this.client.queryContractSmart(this.contractAddress, {
      join_pool: {
        assets,
        auto_stake: autoStake,
        lp_to_mint: lpToMint,
        pool_id: poolId,
        recipient,
        slippage_tolerance: slippageTolerance
      }
    });
  };
  swap = async ({
    recipient,
    swapRequest
  }: {
    recipient?: string;
    swapRequest: SingleSwapRequest;
  }): Promise<SwapResponse> => {
    return this.client.queryContractSmart(this.contractAddress, {
      swap: {
        recipient,
        swap_request: swapRequest
      }
    });
  };
  proposeNewOwner = async ({
    expiresIn,
    owner
  }: {
    expiresIn: number;
    owner: string;
  }): Promise<ProposeNewOwnerResponse> => {
    return this.client.queryContractSmart(this.contractAddress, {
      propose_new_owner: {
        expires_in: expiresIn,
        owner
      }
    });
  };
  dropOwnershipProposal = async (): Promise<DropOwnershipProposalResponse> => {
    return this.client.queryContractSmart(this.contractAddress, {
      drop_ownership_proposal: {}
    });
  };
  claimOwnership = async (): Promise<ClaimOwnershipResponse> => {
    return this.client.queryContractSmart(this.contractAddress, {
      claim_ownership: {}
    });
  };
}"
`;
